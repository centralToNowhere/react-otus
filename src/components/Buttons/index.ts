export { ButtonGameControl } from "./ButtonGameControl";
export { Button } from "./Button";
