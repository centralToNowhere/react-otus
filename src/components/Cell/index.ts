export { Cell } from "./Cell";
export type { ICell, ICellProps } from "./Cell";
