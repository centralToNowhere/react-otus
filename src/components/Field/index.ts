export { Field } from "./Field";
export type { IFiguresMap, IFieldProps } from "./Field";
