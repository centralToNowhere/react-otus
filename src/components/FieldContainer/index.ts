export {
  FieldContainer,
  createFormKey,
  getRandomCells,
  getGameCycleTimeout,
} from "./FieldContainer";
export type { FieldContainerProps } from "./FieldContainer";
