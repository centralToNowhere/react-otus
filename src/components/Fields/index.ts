export { InputField } from "./InputField";
export { LabelField } from "./LabelField";
export { InputPatterns } from "./Properties";
export {
  fieldControlSlice,
  setCellSize,
  setMaxFieldWidth,
  setMaxFieldHeight,
  setCapacity,
  setSpeed,
  resetFieldControls,
  defaultFieldControlState,
} from "./FieldControlRdx";
export type { IFieldControlState } from "./FieldControlRdx";
export { useOnChangeHandler, onBlurHandler } from "./FieldHandlers";
