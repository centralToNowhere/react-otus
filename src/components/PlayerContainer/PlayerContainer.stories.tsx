import React from "react";
import { Story, Meta } from "@storybook/react";
import {
  PlayerContainer,
  IPlayerContainerProps,
} from "@/components/PlayerContainer";

export default {
  title: "Components/PlayerContainer",
  component: PlayerContainer,
  args: {
    player: {
      registered: true,
      name: "Player1",
    },
    onPlayerUnregister: () => {
      //empty
    },
  },
} as Meta;

const Template: Story<IPlayerContainerProps> = (args) => {
  return <PlayerContainer {...args} />;
};

export const PlayerContainerDefault = Template.bind({});
