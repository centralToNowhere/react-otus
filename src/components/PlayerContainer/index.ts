export { PlayerContainer } from "./PlayerContainer";
export type { IPlayerContainerProps } from "./PlayerContainer";
