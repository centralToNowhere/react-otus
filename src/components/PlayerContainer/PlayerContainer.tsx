import React, { FC } from "react";
import { FormElement } from "@/components/Form";
import { css } from "@emotion/react";
import { FormField } from "@/components/Form";
import { FormButton } from "@/components/Buttons";
import { l10n } from "@/l10n/ru";
import { BREAKPOINTS, COLORS } from "@/styles/ui-styled";
import { IPlayer } from "@/player/Player";

export interface IPlayerContainerProps {
  player: IPlayer;
  onPlayerUnregister: () => void;
}

export const PlayerContainer: FC<IPlayerContainerProps> = (props) => {
  return (
    <FormElement
      css={css`
        align-self: flex-start;

        @media screen and (min-width: ${BREAKPOINTS.xl}) {
          position: absolute;
          margin-right: auto;
          left: 0;
        }
      `}
    >
      <FormField
        css={css`
          display: inline-block;

          h1.player-name {
            font-size: 1.2rem;
            font-weight: bold;
            margin: 0;
            border-bottom: 2px solid ${COLORS.accent};
          }

          div.player-stats-container {
            margin: 20px 0;

            & > h2 {
              font-size: 1.2rem;
            }

            & > ul {
              padding-left: 0;
            }
          }

          li {
            margin: 10px 0 10px 20px;
          }
        `}
      >
        <h1 className="player-name">{props.player.name}</h1>
        <div className="player-stats-container">
          <h2>{l10n.statistics}</h2>
          <ul>
            <li>{l10n.score} 0</li>
            <li>{l10n.clicksCount} 0</li>
          </ul>
        </div>
        <FormButton
          css={css`
            text-align: center;
          `}
          type="button"
          onClick={props.onPlayerUnregister}
        >
          {l10n.logoutButton}
        </FormButton>
      </FormField>
    </FormElement>
  );
};
