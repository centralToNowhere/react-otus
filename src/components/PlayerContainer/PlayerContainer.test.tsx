import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import { PlayerContainer } from "@/components/PlayerContainer";
import { l10n } from "@/l10n/ru";
import userEvent from "@testing-library/user-event/dist";

describe("PlayerContainer tests", () => {
  it("should render PlayerContainer", () => {
    jest.spyOn(global.Math, "random").mockReturnValue(0);
    const onPlayerUnregister = jest.fn();
    const { asFragment } = render(
      <PlayerContainer
        onPlayerUnregister={onPlayerUnregister}
        player={{
          registered: true,
          name: "Ivan",
        }}
      />
    );
    jest.spyOn(global.Math, "random").mockRestore();

    const playerNameHeading = screen.getByRole("heading", { level: 1 });
    const statisticsHeading = screen.getByRole("heading", { level: 2 });
    const score = screen.getByText(l10n.score, { exact: false });
    const clicksCount = screen.getByText(l10n.clicksCount, { exact: false });

    expect(playerNameHeading).toHaveTextContent("Ivan");
    expect(statisticsHeading).toHaveTextContent(l10n.statistics);
    expect(score).toBeInTheDocument();
    expect(clicksCount).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should call onPlayerUnregister on player logout", async () => {
    const onPlayerUnregister = jest.fn();

    render(
      <PlayerContainer
        onPlayerUnregister={onPlayerUnregister}
        player={{
          registered: true,
          name: "Ivan",
        }}
      />
    );

    const unregisterButton = screen.getByRole("button");
    userEvent.click(unregisterButton);

    await waitFor(() => {
      expect(onPlayerUnregister).toHaveBeenCalledTimes(1);
    });
  });
});
