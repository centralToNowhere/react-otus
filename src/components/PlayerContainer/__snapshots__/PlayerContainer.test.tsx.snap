// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerContainer tests should render PlayerContainer 1`] = `
<DocumentFragment>
  <form
    class="css-ixoyv7-FormElement-PlayerContainer e1dt6zfi0"
  >
    <div
      class="css-1mkqqr3-FormField-PlayerContainer eybdsu10"
    >
      <h1
        class="player-name"
      >
        Ivan
      </h1>
      <div
        class="player-stats-container"
      >
        <h2>
          Статистика
        </h2>
        <ul>
          <li>
            Очки: 0
          </li>
          <li>
            Количество кликов: 0
          </li>
        </ul>
      </div>
      <button
        class="css-rsh0un-FormButton-PlayerContainer e2n0d7s0"
        type="button"
      >
        Выйти
      </button>
    </div>
  </form>
</DocumentFragment>
`;
